### PR TITLE
Fix flickery unbuffered examples

### PR DESCRIPTION
**Pull request by ctz**
This PR is aiming to fix flickeryness in daily-tests:

- https://github.com/rustls/rustls/actions/runs/9602495561/job/26483466289 - windows stack exhaustion: I don't really understand this, apparently the windows main thread is 1MB but I neither understand why this isn't enough, and why it isn't grown on a fault? It seems to only affect the `unbuffered-async-client` example, but I've fixed both for completeness.
- https://github.com/rustls/rustls/actions/runs/9570253282/job/26384594746 - the unbuffered-*-client tests are flickery in certain network conditions, explained the commit message.

### CODE REVIEW

**Comment by ctz** [comment]
Example run: https://github.com/rustls/rustls/actions/runs/9616685499

File: rustls/tests/api.rs [hunk 1]
```diff
@@ -5344,6 +5344,14 @@ fn connection_types_are_not_huge() {
...
     assert_lt(mem::size_of::<ServerConnection>(), 1600);
     assert_lt(mem::size_of::<ClientConnection>(), 1600);
+    assert_lt(
+        mem::size_of::<rustls::server::UnbufferedServerConnection>(),
+        1600,
+    );
+    assert_lt(
+        mem::size_of::<rustls::client::UnbufferedClientConnection>(),
+        1600,
+    );
```
Comment (cpu) [inline rustls/tests/api.rs:5354]: small nit: I think this change is unrelated to the stack exhaustion fix (though still a good idea!). Would you be willing to pull it out into a separate commit?

**Review by cpu — APPROVED** [review]
Thank you :bow:
